examples/server/api_like_OAI.py

```diff
@@ -46,18 +50,184 @@ def convert_chat(messages):
             prompt += f"{system_n}{line['content']}"
         if (line["role"] == "user"):
             prompt += f"{user_n}{line['content']}"
+        if (line["role"] == "function"):
+            prompt += f"{fn_n}{line['content']}"
         if (line["role"] == "assistant"):
             prompt += f"{ai_n}{line['content']}{stop}"
     prompt += ai_n.rstrip()
 
     return prompt
 
-def make_postData(body, chat=False, stream=False):
+def make_grammar(schema, root):
+    indent_inc = "  "
+    def format_rulename(name):
+        return name.replace("_", "-")
+
+    def is_basetype(typename):
+        return typename in ['integer', 'number', 'bool', 'string']
+
+    def schema_typename(prefix, schema, defs, arrs):
+        typename = '"null"'
+        if 'type' in schema:
+            typename = schema['type']
+            if not is_basetype(typename) and typename != 'array':
+                typename = prefix + typename
+        if '$ref' in schema and schema['$ref'] in defs:
+            typename = defs[schema['$ref']]
+        if typename == 'array':
+            elemtype = schema_typename(prefix, schema['items'], defs, arrs)
+            typename = f'array-{elemtype}'
+            if not is_basetype(elemtype):
+                typename = prefix + typename
+            typename = format_rulename(typename)
+            if typename not in arrs:
+                arrs[typename] = elemtype
+        return typename
+
+    def arr_to_rules(rules, prefix, name, elemtype):
```

Comment:
Is there a reason arrays are treated specially, rather than a recursive case along with objects? Also, this approach seems to miss the element definition in the case of non-primitive elements:
```
>>> print(make_grammar([{'name': 'foo', 'parameters': {'type': 'object', 'properties': {'a': {'type': 'array', 'items': {'type': 'object', 'properties': {'a': {'type': 'number'}}}}}}}], {'name': 'foo'}))
foo ::= "{" "\"" "a" "\"" ":" foo-array-foo-object "}"
foo-array-foo-object ::= "[" ( foo-object ( "," foo-object )* )? "]"
root ::= foo
ws ::= [ \t\n]?
string ::=
  "\"" (
    [^"\\] |
    "\\" (["\\/bfnrt] | "u" [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F]) # escapes
  )* "\""
bool ::= "True" | "False"
integer ::= ("-"? ([0-9] | [1-9] [0-9]*))
number ::= ("-"? ([0-9] | [1-9] [0-9]*)) ("." [0-9]+)? ([eE] [-+]? [0-9]+)?
```